### PR TITLE
Connector state (fix races)

### PR DESCRIPTION
**Pull request by lovromazgon**
### Description

We want to track the connector state to prevent race conditions when executing async functions, especially in `Stop`.

The connector state is now tracked _kind of_ like in a state machine. Every connector function executes code while holding the lock in the newly introduced `ConnectorStateWatcher`. It's essentially a `ValueWatcher` that adds on specific functionality around connector states.

### Quick checks:

- [ ] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [ ] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-sdk/pulls) for the same
  update/change.
- [ ] I have written unit tests.
- [ ] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

**Review by maha-hajja — APPROVED** [review]
lovely work with the connector state! 
these are mini nitpicks, otherwise looks good, just let me know when you finish the tests :D

File: destination.go [hunk 1]
```diff
@@ -233,13 +233,18 @@ func (a *destinationPluginAdapter) Stop(ctx context.Context, req cpluginv1.Desti
...
+	// to flush what we have s far
+	_, err := a.lastPosition.Watch(
+		ctx,
+		func(val Position) bool {
+			return bytes.Equal(val, req.LastPosition)
+		},
+		csync.WithTimeout(stopTimeout),
+	)
 
 	// flush cached records, allow it to take at most 1 minute
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply destination.go:246]: Yes, I just attached it to the same constant [here](https://github.com/ConduitIO/conduit-connector-sdk/pull/89/files/62809a547b3b7c7a4ec4784fa3866e5bf3c4df9f#diff-a4d006ab5233eeae48e5dc55c19a888b1a1ab4edcda548ba466e8a88909ed063R37).

[the remark this answers is not in the document]
Comment (maha-hajja) [reply destination.go:246]: oh nice, so this comment should be updated cuz it's 2 minutes now

[the remark this answers is not in the document]
Comment (lovromazgon) [reply destination.go:246]: Hmm, flushing is still constrained to 1 minute, you probably mean the whole stop operation, which can take 2 minutes in the worst case (1 minute for waiting for the last record, 1 minute for flush). So I think this comment is fine, but I can add another one on top of `Stop` to make it clear what we can expect.

[the remark this answers is not in the document]
Comment (lovromazgon) [reply destination.go:258]: wdyt about this https://github.com/ConduitIO/conduit-connector-sdk/pull/89/commits/adc48cc4eddb067bfffa48f82df19db234437981?

**Pull request merged**